Add RAKWireless RAK11300 Module

ScholliYT opened the pull request:
This PR adds support for the [RAK11300](https://store.rakwireless.com/products/wisduo-lpwan-module-rak11300) Module to RadioBoards to make it easier to use with RadioLib.

The code is mostly based on https://www.nico-maas.de/?p=2607 from @nmaas87

Features:
* Add support for RAK11300 with SX1262
* Use SPI1 as defined in Arduino-Pico
* Use custom RF Switch

TODO:
- [x] Auto detect board based on compiler flags (any flag that can we use?)

ScholliYT committed:
Add RAKWireless WisDuo RAK11300 Module

* Add support for RAK11300 with SX1262
* Use SPI1 as definded in Arduino-Pico
* Use custom RF Switch
* TODO: Auto detect board based on which flag?

ScholliYT committed:
fix name of RAK11300

jgromes commented on src/contributed/RAKWireless/RAK11300.h:
```diff
@@ -0,0 +1,50 @@
+
+  #define RADIO_RF_SWITCH_TABLE \
+  static const Module::RfSwitchMode_t RadioBoards_rfswitch_table[] = {  \
+    /* mode              ANT_PWR */                         \
+    {Module::MODE_IDLE,  {LOW}},                            \
+    {Module::MODE_RX,    {HIGH}},                           \
+    {Module::MODE_TX,    {HIGH}},                           \
+    {Module::MODE_IDLE,  {LOW}},                            \
```
Why is idle mode twice here? It should be enough to have it just once.

ScholliYT replied:
You're absolutely right. Stupid copy paste error...

jgromes commented:
Thank you for the contribution! I left one comment. 

Do you want to keep this open until you figure out the automated detection? If so, then please mark it as draft so that it doesn't get merged prematurely.

ScholliYT committed:
fix duplicate state in rfswitch table

serialrf433 commented:
I also had some struggle to understand how the automatic detection is working.
Here is the solution for esp32: https://github.com/radiolib-org/RadioBoards/issues/5#issuecomment-2721992913

But the RAK1130 is not esp32. That is why you have to take a look at the arduino library you are using to add support for this board when using the library you have chosen. If you for example use this: https://github.com/RAKWireless/RAK-RP-Arduino
Then inside here https://raw.githubusercontent.com/RAKWireless/RAK-RP-Arduino/refs/heads/main/boards.txt its the line WisCoreRAK11300Board.build.board=RASPBERRY_PI_PICO

So the name for autodetection when using this library would be: **RASPBERRY_PI_PICO**

I hope you are using some other library, because RAK's choice is not that great to be added to RadioBoards because its far too general.

StevenCellist commented:
Compiling a random sketch for the RAK11300 on the Arduino IDE with verbose compilation uncovers the following definition:

```cpp
ARDUINO_RAKWIRELESS_RAK11300
```

This RAK11300 is present in the unofficial RP2040 core - it is not the BSP from RAKWireless linked above by @serialrf433. So the question would be which core @ScholliYT wishes to use. Personally, I'd say using the RP2040 core makes sense.

ScholliYT commented:
I was wondering about the same thing. Which BSP should be chosen if any. 

Personally, I am using PlatformIO with the following setup based on a [fork](https://github.com/maxgerhardt/platform-raspberrypi.) of the unofficial [RP2040 core](https://github.com/earlephilhower/arduino-pico).
```ini
[env:rakwireless_rak11300]
platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board = rakwireless_rak11300
framework = arduino
board_build.core = earlephilhower
monitor_speed = 115200
lib_deps = 
	jgromes/RadioLib@^7.1.2
	; jgromes/RadioBoards@^1.0.0
        git+git@github.com:ScholliYT/RadioBoards.git#rakwireless-rak11300
```

We have `ARDUINO_RAKWIRELESS_RAK11300` in [there](https://github.com/maxgerhardt/platform-raspberrypi/blob/daa66b4a9853aae38bdcc482f0a702215013425b/boards/rakwireless_rak11300.json#L12) as @StevenCellist commented.

StevenCellist commented:
Definitely use the RP2040, I'd say - it's actually maintained. So let's stick to `ARDUINO_RAKWIRELESS_RAK11300`. Could you kindly add it to cross off your TODO?

ScholliYT committed:
automatically resolve RAK11300

The flag ARDUINO_RAKWIRELESS_RAK11300 is defined in the 
unofficial Arduino Pico RP2040 BSP which we recommend to use.

ScholliYT commented:
Done. This should be good to go now @jgromes

jgromes commented:
@ScholliYT merged - thank you for the contribution!
